### PR TITLE
Phys

### PR DESCRIPTION
**Pull request by iarba**

### CODE REVIEW

File: src/chat_server/shutdown.h [hunk 1]
```diff
@@ -0,0 +1,12 @@
+#ifndef CHAT_SERVER_SHUTDOWN_H
+#define CHAT_SERVER_SHUTDOWN_H
+
+#include <boost/thread/barrier.hpp>
+
+void shutdown();
+
+void shutdown(int _whatever);
```
Comment (AiMikoto) [inline src/chat_server/shutdown.h]: weird name choice for a function parameter, but... _whatever

File: src/components/collider.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
+#include "components/collider.h"
+#include <algorithm>
+
+collider::collider(glm::dvec3 size, collider_t c_type) : Component(false)
```
Comment (AiMikoto) [inline src/components/collider.cpp]: A GameObject can, and should have multiple colliders. For example: one that can be a trigger so the player can interact with the object from far away, and one that is just bounding volume itself, so the player, or other entities, can't just walk through it.

File: src/components/physical.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  void initialise();
+  void tick(double delta);
+  void add_force(glm::dvec3 force);
+  glm::dvec3 velocity;
+  glm::dvec3 force_acc;
+  double m;
+  double im;
+  bool collides;
+  bool movable;
+  bool collidable;
```
Comment (AiMikoto) [inline src/components/physical.h]: A game object with a physical component attached to it always moves, meaning we will always apply gravity to it, so the bool is redundant. Basically, the component's presence specify that the object is movable. Also you might want to keep track of collision presence within the collider component, as that is what has the bounds defined, and not the physical one.

File: src/chat_server/ioc.h [hunk 4]
```diff
@@ -1,8 +1,4 @@
 #ifndef CHAT_SERVER_IOC_H
 #define CHAT_SERVER_IOC_H
```
Comment (AiMikoto) [inline src/chat_server/ioc.h]: what is the purpose of this file?

File: src/chat_server/ioc.h [hunk 5]
```diff
@@ -1,8 +1,4 @@
...
 
-#include <boost/asio.hpp>
-
-extern boost::asio::io_context ioc;
-
 #endif // CHAT_SERVER_IOC_H
```
Comment (AiMikoto) [inline src/chat_server/ioc.h]: What is the purpose of this file?

File: src/instance/ioc.h [hunk 6]
```diff
@@ -1,10 +1,8 @@
 #ifndef SERVER_IOC_H
 #define SERVER_IOC_H
 
-#include <boost/asio.hpp>
 #include "lib/database.h"
 
-extern boost::asio::io_context ioc;
 extern database *db;
```
Comment (AiMikoto) [inline src/instance/ioc.h]: You  declare the same external with the same name in server/ioc.h . Might want to watch out for that. Are the server and instances database different from one another? You might consider using namespaces for server and instance to keep them separated if that's the case.

File: src/lib/call.cpp [hunk 7]
```diff
@@ -35,7 +39,17 @@ call read_call(boost::asio::ip::tcp::socket *s, crypto *cry)
...
   uint32_t length = *(uint32_t *)(make_string(read_buffer).c_str());
+  while(s -> available() < length)
+  {
+    trials++;
+    boost::this_thread::sleep_for(boost::chrono::milliseconds(1));
+    if(trials == MAX_TRIAL_SEQUENT)
+    {
+      throw std::logic_error("too slow connection");
+    }
+  }
```
Comment (AiMikoto) [inline src/lib/call.cpp]: am not sure what this is supposed to check, or protect against

File: src/lib/rcon.cpp [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+#include "lib/log.h"
+
+boost::asio::io_context rcon_ioc;
+
+rcon::rcon(std::string hostname, int port, std::string tok, rsa_crypto *rsa, aes_crypto *aes)
+{
+  boost::asio::ip::tcp::socket *sock = new boost::asio::ip::tcp::socket(rcon_ioc);
+  BOOST_LOG_TRIVIAL(info) << "attempting rcon to " << hostname << ":" << port;
+  boost::asio::ip::tcp::resolver resolver(rcon_ioc);
+  boost::asio::ip::tcp::resolver::results_type endpoint = resolver.resolve(boost::asio::ip::tcp::v4(), hostname, std::to_string(port));
```
Comment (AiMikoto) [inline src/lib/rcon.cpp]: Unless I'm missing something, it looks like you're ignoring users with ipv6 adresses?

File: src/phys/aabb.cpp [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+
+aabb root_aabb()
+{
+  aabb ret;
+  ret.minx = -1000;
+  ret.miny = -1000;
+  ret.minz = -1000;
+  ret.maxx =  1000;
+  ret.maxy =  1000;
+  ret.maxz =  1000;
```
Comment (AiMikoto) [inline src/phys/aabb.cpp]: Later on, we might want not to have this values hardcoded, just in case we have scenes larger than 2000 x 2000 x 2000

File: src/phys/octree.cpp [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+      {
+        children[index] -> insert(it.first, it.second);
+      }
+      lazy_boxes[index].clear();
+      std::set<int> difference;
+      std::set_difference(children[index] -> boxes_i.begin(), children[index] -> boxes_i.end(), ret -> begin(), ret -> end(), std::inserter(difference, difference.end()));
+      if(difference.size() != 0)
+      {
+	children[index] -> get_collisions_h(box, ret);
+      }
```
Comment (AiMikoto) [inline src/phys/octree.cpp]: Can't this be placed in a for? It's hard to follow + it seems to be the same code for all indices...

File: src/server/main.cpp [hunk 11]
```diff
@@ -7,16 +7,28 @@
...
 #include <boost/thread/barrier.hpp>
+#include "server/shutdown.h"
+#include "server/hosts.h"
+#include "server/token.h"
+
+#ifdef __linux__
+#include <csignal>
+#endif
+
+std::string my_token = "rubber";
```
Comment (AiMikoto) [inline src/server/main.cpp]: rubberbanding Pog

File: src/chat_server/shutdown.h (hunk 1 above)
Comment (iarba) [reply src/chat_server/shutdown.h]: shutdown is passed as a callback to signal handling, and it must take an int, which i don't use as u can see

File: src/components/collider.cpp (hunk 2 above)
Comment (iarba) [reply src/components/collider.cpp]: will later switch to inheritance, where each collider is actually unique in type

File: src/components/physical.h (hunk 3 above)
Comment (iarba) [reply src/components/physical.h]: > A game object with a physical component attached to it always moves, meaning we will always apply gravity to it
surfaces are not movable, but they are collidable. think about it

File: src/chat_server/ioc.h (hunk 5 above)
Comment (iarba) [reply src/chat_server/ioc.h]: obsolete, can delete

File: src/instance/ioc.h (hunk 6 above)
Comment (iarba) [reply src/instance/ioc.h]: they are in different executables

File: src/lib/call.cpp (hunk 7 above)
Comment (iarba) [reply src/lib/call.cpp]: let's say i have a server.

i have a client 1 connecting to the server, allocating memory. but client 1 is malicious and never writes anything.

client 2 is also malicious and does the same.

...

client n is also malicious and does the same.

server becomes overloaded.

this check is set in place to make sure that the client actually sends data, protecting against DOS attack

File: src/lib/rcon.cpp (hunk 8 above)
Comment (iarba) [reply src/lib/rcon.cpp]: true, but not a priority, can be handled later

File: src/phys/aabb.cpp (hunk 9 above)
Comment (iarba) [reply src/phys/aabb.cpp]: you can initialise the octree with an aabb of custum size.

File: src/phys/octree.cpp (hunk 10 above)
Comment (iarba) [reply src/phys/octree.cpp]: i couldn't find a more elegant way to solve this, if you can think of one, go ahead. there are unit tests, you can develop against those

File: src/server/main.cpp (hunk 11 above)
Comment (iarba) [reply src/server/main.cpp]: you should probably sleep

File: src/components/physical.h (hunk 3 above)
Comment (AiMikoto) [reply src/components/physical.h]: Yeah but the physical component should only apply forces to the object. Meaning, if it's not present, an object should not be able to move, as it wouldn't get forces applied to it. Also, just recalled, these are called rigid bodies in physics, might wanna work with that name, for clarity. Remember a surface or terrain will only have a collider attached to it, not a rigid body. Same for multiple other game objects that we want to define as "static" such as trees or not. Otherwise, we'd just have to add one to each object in the scene and worry about setting this property. Which beats the point of game objects have their behavior be defined by the presence of a component.

File: src/components/collider.cpp (hunk 2 above)
Comment (AiMikoto) [reply src/components/collider.cpp]: I'm not sure what you mean by unique in type, but even in that case we could still have multiple colliders of the same type. For instance, think about 2 sphere triggers, one bigger than the other. You could define a behaviour for when an object is close, and a separate one for when it's very close. As a quick example imagine a minigame of king of the hill or capture the point something like that. You can have that players that are on the edges of the point contribute less than those that are at the center and have less room to maneuver.

File: src/components/physical.h (hunk 3 above)
Comment (iarba) [reply src/components/physical.h]: we should talk in detail about this
